doc/man/mutter-message.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.\" -----
.\" This file was confirmed to be licenced under the GPL
.\" by its author and copyright holder, Akira TAGOH, on June 1st 2008:
.\"
.\" > I'm comfortable with DFSG-free. that sounds great if you think it's
.\" > useful and worth containing it in upstream.
.\" ...
.\" > Right I know. any licenses that is DFSG-free, I'm ok with whatever,
.\" > since I have contributed that for Debian. so GPL is no problem for me.
.\" -----
.TH MUTTER\-MESSAGE 1 "28 August 2002"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
MUTTER\-MESSAGE \- a command to send a message to Mutter
.SH SYNOPSIS
.B MUTTER\-MESSAGE
[restart|reload\-theme|enable\-keybindings|disable\-keybindings]
.SH DESCRIPTION
This manual page documents briefly the
.B mutter\-message\fP.
This manual page was written for the Debian distribution
because the original program does not have a manual page.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBmutter\-message\fP send a specified message to \fBmutter\fP(1).
.SH OPTIONS
.TP
.B restart
Restart \fBmutter\fP(1) which is running.
.TP
.B reload-theme
Reload a theme which is specified on gsettings database.
.TP
.B enable-keybindings
Enable all of keybindings which is specified on gsettings database.
.TP
.B disable-keybindings
Disable all of keybindings which is specified on gsettings database.
.SH SEE ALSO
.BR mutter (1)
.SH AUTHOR
This manual page was written by Akira TAGOH <tagoh@debian.org>,
for the Debian GNU/Linux system (but may be used by others).


NEWS
Cogl 1.7.2                                                         2011-07-01
===============================================================================

  • List of changes since Clutter 1.7.0

   » Add an example program to draw a 3D crate

   » Add experimental API to get and set the viewport on a framebuffer
     without going through the implicit context.

   » Add API to select the winsys backend without having to use an
     environment variable.

   » Various documentation fixes.

   » Add winsys-specific typesafe APIs to replace the old
     cogl_renderer_handle_native_event functions.

   » Rename platform-specific APIs so that we consistently have
     cogl_<platform>_symbol_name instead of cogl_symbol_name_<platform>.

   » Fixed building against GLES.

  • List of bugs fixes since Clutter 1.7.0

    https://bugzilla.gnome.org:

    #652672 - Typo fix in cogl-framebuffer
    #653615 - cogl_release_status='snapshot' breaks linking on 64-bit
              systems due to namespace conflict with ltoptions.m4
    #653229 - pipeline: Don't regenerate shaders when not changing the
              number of layers
    #653231 - pipeline: Don't trigger a change when binding the invalid
              texture target

Many thanks to:
  Robert Bragg
  Adam Jackson
  Emmanuele Bassi

Cogl 1.7.0                                                         2011-06-10
===============================================================================

  • List of changes since Clutter 1.6

    » Cogl split out into a standalone project!

    » Added native winsys backends for GLX, WGL, EGL and Android (Quartz in
      progress)
      » Supported EGL platforms include: NULL, X11, GDL and Wayland

    » Adds renderer,display,onscreen-template and swap-chain APIs for
      writing standalone apps using the experimental 2.0 API

    » Added convenience functions cogl_matrix_view_2d_in_frustum/perspective
      for setting up a Clutter style view transform for a 2d coordinate system
      inside a projective perspective.

    » Added an experimental Quaternion API

    » Added an experimental Euler API

    » Re-worked the CoglPipeline depth testing state API

    » Added EGL X11 texture-from-pixmap support.

    » Added a tiny Wayland compositor (without input support) under
      examples/

    » Optimizations:
      » Optimise paths that are just a rectangle
      » Journal: Use a pool of vertex-arrays to avoid allocation costs.

  • List of bugs fixes since Clutter 1.6

    http://bugzilla.clutter-project.org:
    #2538 - crash in gnome-shell
    #2544 - Using vertex buffers causes memory corruption when you unref the CoglHandle
    #2561 - Updates to COGL sources regarding non-standard code usage
    #2588 - Assertion failure in _cogl_pipeline_set_layer_texture_target
    #2594 - cogl-clip-state: Adapt to experimental cogl2 API.
    #2620 - _cogl_offscreen_new_to_texture_full() leaks reference when failing

    https://bugzilla.gnome.org:
    #641197 - mouse events at pixel row zero are ignored
    #650966 - CoglPipeline doesn't support any layers on fixed function hardware
    #650979 - cogl-pipeline-fragend-fixed: Fix enabling texture targets
    #651256 - configure fails for the poor sods who use dash as sh

Many thanks to:

  Neil Roberts
  Emmanuele Bassi
  Damien Lespiau
  James Athey
  Jasper St. Pierre
  Piotr Drąg


src/tools/mutter-message.c
/* Mutter send-magic-messages app */

/*
 * Copyright (C) 2002 Havoc Pennington
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 */

#include <config.h>
#include <gtk/gtk.h>
#include <gdk/gdkx.h>
#include <stdlib.h>
#include <string.h>

#include <libintl.h>
#define _(x) dgettext (GETTEXT_PACKAGE, x)
#define N_(x) x

static Display *display;

static void
send_reload_theme (void)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = display;
  xev.xclient.window = gdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (display,
                                          "_MUTTER_RELOAD_THEME_MESSAGE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = 0;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (display,
              gdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (display);
  XSync (display, False);
}

static void
send_set_keybindings (gboolean enabled)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = display;
  xev.xclient.window = gdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (display,
                                          "_MUTTER_SET_KEYBINDINGS_MESSAGE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = enabled;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (display,
              gdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (display);
  XSync (display, False);
}

#ifdef WITH_VERBOSE_MODE
static void
send_toggle_verbose (void)
{
  XEvent xev;

  xev.xclient.type = ClientMessage;
  xev.xclient.serial = 0;
  xev.xclient.send_event = True;
  xev.xclient.display = display;
  xev.xclient.window = gdk_x11_get_default_root_xwindow ();
  xev.xclient.message_type = XInternAtom (display,
                                          "_MUTTER_TOGGLE_VERBOSE",
                                          False);
  xev.xclient.format = 32;
  xev.xclient.data.l[0] = 0;
  xev.xclient.data.l[1] = 0;
  xev.xclient.data.l[2] = 0;

  XSendEvent (display,
              gdk_x11_get_default_root_xwindow (),
              False,
	      SubstructureRedirectMask | SubstructureNotifyMask,
	      &xev);

  XFlush (display);
  XSync (display, False);
}
#endif

static void
usage (void)
{
  g_printerr (_("Usage: %s\n"),
              "mutter-message (reload-theme|enable-keybindings|disable-keybindings|toggle-verbose)");
  exit (1);
}

int
main (int argc, char **argv)
{
  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
  
  gtk_init (&argc, &argv);

  if (argc < 2)
    usage ();

  display = GDK_DISPLAY_XDISPLAY (gdk_display_get_default ());

  if (strcmp (argv[1], "reload-theme") == 0)
    send_reload_theme ();
  else if (strcmp (argv[1], "enable-keybindings") == 0)
    send_set_keybindings (TRUE);
  else if (strcmp (argv[1], "disable-keybindings") == 0)
    send_set_keybindings (FALSE);
  else if (strcmp (argv[1], "toggle-verbose") == 0)
    {
#ifndef WITH_VERBOSE_MODE
      g_printerr (_("Mutter was compiled without support for verbose mode\n"));
      return 1;
#else      
      send_toggle_verbose ();
#endif
    }
  else
    usage ();
  
  return 0;
}

